rust/extns/numaflow-kafka/src/lib.rs

```diff
@@ -0,0 +1,680 @@
+            topic: config.topic,
+            handler_rx,
+        };
+
+        actor
+            .pending_messages()
+            .await
+            .map_err(|err| Error::Kafka(format!("Failed to get pending messages: {err:?}")))?;
+
+        tokio::spawn(async move {
+            tracing::info!("Starting Kafka consumer...");
+            actor.run().await;
+        });
+
+        Ok(())
+    }
+
+    async fn run(&mut self) {
+        while let Some(msg) = self.handler_rx.recv().await {
+            self.handle_message(msg).await;
+        }
+    }
+
+    async fn handle_message(&mut self, msg: KafkaActorMessage) {
+        match msg {
+            KafkaActorMessage::Read { respond_to } => {
+                let messages = self.read_messages().await;
+                let _ = respond_to.send(messages);
+            }
+            KafkaActorMessage::Ack {
+                offsets,
+                respond_to,
+            } => {
+                let status = self.ack_messages(offsets).await;
+                let _ = respond_to.send(status);
+            }
+            KafkaActorMessage::Pending { respond_to } => {
+                let pending = self.pending_messages().await;
+                let _ = respond_to.send(pending);
+            }
+        }
+    }
+
+    async fn read_messages(&mut self) -> Result<Vec<KafkaMessage>> {
+        let mut messages: Vec<KafkaMessage> = vec![];
+        let timeout = tokio::time::timeout(self.read_timeout, std::future::pending::<()>());
+        tokio::pin!(timeout);
+        loop {
+            if messages.len() >= self.batch_size {
+                break;
+            }
+            tokio::select! {
+                biased;
+
+                _ = &mut timeout => {
+                    break;
+                }
+
+                message = self.consumer.recv() => {
+                    let message = match message {
+                        Ok(msg) => msg,
+                        Err(e) => {
+                            error!(?e, "Failed to read messages, will retry after 1 second");
+                            tokio::time::sleep(Duration::from_secs(1)).await;
+                            continue;
+                        }
+                    };
+
+                    let headers = match message.headers() {
+                        Some(headers) => headers
+                            .iter()
+                            .map(|header| {
+                                (
+                                    header.key.to_string(),
+                                    String::from_utf8_lossy(header.value.unwrap_or_default()).to_string(),
+                                )
+                            })
+                            .collect(),
+                        None => HashMap::new(),
+                    };
+
+                    let value = match message.payload() {
+                        Some(payload) => Bytes::copy_from_slice(payload),
+                        None => Bytes::new(),
```

Comment:
when will this be `None`, empty payload?

Reply:
Added comment in the code.
```rust
// The rdkafka doc says that the payload can be None if there is no payload.
```